### PR TITLE
feat(module:tree, tree-select): support customize icons

### PR DESCRIPTION
**Pull request by danranVm**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/NG-ZORRO/ng-zorro-antd/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

```
[ ] Bugfix
[x] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Documentation content changes
[ ] Application (the showcase website) / infrastructure changes
[ ] Other... Please describe:
```

## What is the current behavior?

Issue Number: #2890 #2503

## Does this PR introduce a breaking change?
```
[ ] Yes
[x] No
```

### CODE REVIEW

File: components/tree/nz-tree.component.ts [hunk 1]
```diff
@@ -34,10 +34,11 @@ import { NzTreeService } from './nz-tree.service';
 
 export class NzTreeComponent implements OnInit, OnChanges, OnDestroy {
   @Input() @InputBoolean() nzShowIcon = false;
+  @Input() @InputBoolean() nzShowExpand = true;
   @Input() @InputBoolean() nzShowLine = false;
+  @Input() nzExpandedIcon: string | TemplateRef<void>;
```
Comment (HsuanXyz) [inline components/tree/nz-tree.component.ts]:
```ts
@Input() nzExpandedIcon: string | TemplateRef<NzTreeNode>;
```

File: components/tree/nz-tree-node.component.html [hunk 2]
```diff
@@ -16,7 +16,11 @@
       [class.ant-tree-switcher_close]="isSwitcherClose"
       (click)="_clickExpand($event)">
       <ng-container *ngIf="isShowSwitchIcon">
-        <i *ngIf="!nzTreeNode.isLoading" nz-icon type="caret-down" class="ant-tree-switcher-icon"></i>
+        <ng-container *ngIf="!nzTreeNode.isLoading">
+          <ng-container *nzStringTemplateOutlet="nzExpandedIcon">
+            <i nz-icon [type]="nzExpandedIcon || 'caret-down'" class="ant-tree-switcher-icon"></i>
```
Comment (HsuanXyz) [inline components/tree/nz-tree-node.component.html]: 这里需要向 `ng-template` 注入 `context`, 因为可能会根据节点属性改变 icon

Comment (danranVm) [reply components/tree/nz-tree-node.component.html]: done

File: components/tree/demo/customized-icon.ts [hunk 3]
```diff
@@ -5,7 +5,16 @@ import { Component, OnInit } from '@angular/core';
...
-      nzShowIcon="true">
+      nzShowIcon="true"
+      [nzExpandedIcon]="'down'">
+    </nz-tree>
+    <nz-tree
+      [nzData]="nodes"
+      nzShowIcon="true"
+      [nzExpandedIcon]='expandedIcon'>
+      <ng-template #expandedIcon let-node>
+        <i nz-icon [type]="node.origin.icon" class="ant-tree-switcher-icon"></i>
```
Comment (simplejason) [inline components/tree/demo/customized-icon.ts]: nzExpandedIcon supports TemplateRef only is better, remove string type, and should consider the nzShowLine case.

**Comment by simplejason** [comment]
抽时间考虑一下string支持问题，如果支持要么取消旋转class，要么对用户说明旋转属于设计问题，之后rebase一下master

**Comment by danranVm** [comment]
@simplejason 移除了 string 的支持，新增了 show line 时也支持自定义图标。

File: components/tree-select/nz-tree-select.component.html [hunk 4]
```diff
@@ -42,6 +42,7 @@
       [nzAsyncData]="nzAsyncData"
       [nzShowExpand]="nzShowExpand"
       [nzShowLine]="nzShowLine"
+      [nzExpandedIcon]='nzExpandedIcon'
```
Comment (wenqi73) [inline components/tree-select/nz-tree-select.component.html]: Please use double quotation marks uniformly.

Comment (danranVm) [reply components/tree-select/nz-tree-select.component.html]: done

**Comment by danranVm** [comment]
@HsuanXyz @simplejason 没其他问题的话，麻烦 merge 一下吧，不然隔段时间就要 rebase

File: components/tree/doc/index.en-US.md [hunk 5]
```diff
@@ -18,6 +18,7 @@ Almost anything can be represented in a tree structure. Examples include directo
 | `[nzCheckable]` | Adds a Checkbox before the treeNodes| `boolean` | `false` |
 | `[nzShowExpand]` | Show a Expand Icon before the treeNodes | `boolean` | `true` |
 | `[nzShowLine]` | Shows a connecting line | `boolean` | `false` |
+| `[nzExpandedIcon]` | Customize a Expand Icon | `TemplateRef<{ $implicit: NzTreeNode }>` | - |
```
Comment (simplejason) [inline components/tree/doc/index.en-US.md]: Customize  an expand icon

File: components/tree/demo/customized-icon.ts [hunk 6]
```diff
@@ -3,7 +3,22 @@ import { Component, OnInit } from '@angular/core';
 @Component({
   selector: 'nz-demo-tree-customized-icon',
   template: `
-    <nz-tree [nzData]="nodes" nzShowIcon> </nz-tree>
+    <nz-tree [nzData]="nodes" nzShowIcon [nzExpandedIcon]="expandedIcon1">
```
Comment (simplejason) [inline components/tree/demo/customized-icon.ts]: expandedIcon1 ->expandedIconTpl  (semantic variable definition is better)

File: components/tree/demo/customized-icon.ts [hunk 7]
```diff
@@ -3,7 +3,22 @@ import { Component, OnInit } from '@angular/core';
...
+      <ng-template #expandedIcon1 let-node>
+        <i nz-icon [type]="node.origin.icon" class="ant-tree-switcher-icon"></i>
+      </ng-template>
+    </nz-tree>
+    <nz-tree [nzData]="nodes" nzShowIcon [nzExpandedIcon]="expandedIcon2">
```
Comment (simplejason) [inline components/tree/demo/customized-icon.ts]: expandedIcon2 - mutiExpandedIconTpl  (semantic variable definition is better)

File: components/tree/demo/customized-icon.ts [hunk 8]
```diff
@@ -3,7 +3,22 @@ import { Component, OnInit } from '@angular/core';
...
+    </nz-tree>
+    <nz-tree [nzData]="nodes" nzShowIcon [nzExpandedIcon]="mutiExpandedIconTpl">
+      <ng-template #mutiExpandedIconTpl let-node>
+        <i
+          *ngIf="!node.origin.isLeaf"
+          nz-icon
+          [type]="node.isExpanded ? 'folder-open' : 'folder'"
+          class="ant-tree-switcher-line-icon"
+        ></i>
+        <i *ngIf="node.origin.isLeaf" nz-icon type="file" class="ant-tree-switcher-line-icon"></i>
```
Comment (wenqi73) [inline components/tree/demo/customized-icon.ts]: @danranVm @simplejason demo 要改一下，`nzExpandedIcon` 就是 非叶子节点才有的。。会造成误解

**Pull request merged**